storage/series/reader.go

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package series
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3db/context"
+	"github.com/m3db/m3db/retention"
+	"github.com/m3db/m3db/storage/block"
+	"github.com/m3db/m3db/ts"
+	xio "github.com/m3db/m3db/x/io"
+	xerrors "github.com/m3db/m3x/errors"
+)
+
+var (
+	errSeriesReadInvalidRange = errors.New(
+		"series invalid time range read argument specified")
+)
+
+// Reader reads results from a series, or a series block
+// retriever or both.
+// It is implemented as a struct so it can be allocated on
+// the stack.
+type Reader struct {
+	opts        Options
+	cloneableID ts.ID
+	retriever   QueryableBlockRetriever
+}
+
+// NewReaderForRetriever returns a reader for a series
+// block retriever, it will only return data from the block
+// retriever.
+func NewReaderForRetriever(
+	id ts.ID,
+	retriever QueryableBlockRetriever,
+	opts Options,
+) Reader {
+	return Reader{opts: opts, cloneableID: id, retriever: retriever}
+}
+
+// ReadEncoded reads encoded blocks using just a block retriever.
+func (r Reader) ReadEncoded(
+	ctx context.Context,
+	start, end time.Time,
+) ([][]xio.SegmentReader, error) {
+	return r.readersWithBlocksMapAndBuffer(ctx, start, end, nil, nil)
+}
+
+func (r Reader) readersWithBlocksMapAndBuffer(
+	ctx context.Context,
+	start, end time.Time,
+	seriesBlocks block.DatabaseSeriesBlocks,
+	seriesBuffer databaseBuffer,
+) ([][]xio.SegmentReader, error) {
+	// TODO(r): pool these results arrays
```

Comment:
Is this TODO a production blocker?

Reply:
Nope, we live with this today in production (ha).  Once we start doing 1MM read / sec we probably will want to do things like this.